In [171]:
import pandas as pd
import numpy as np
import random
import warnings
warnings.filterwarnings("ignore")
from IPython.core.debugger import set_trace
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import Imputer
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn import preprocessing
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn import metrics
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import RidgeCV
from sklearn.linear_model import LassoCV
from sklearn.metrics import mean_squared_error, r2_score

In [172]:
autoIns = pd.read_csv("training_data.csv")

In [173]:
autoIns.shape

(424431, 69)

In [84]:
autoIns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 424431 entries, 0 to 424430
Data columns (total 69 columns):
PolicyNo                                    424431 non-null int64
Policy_Company                              424431 non-null object
Policy_Installment_Term                     424431 non-null int64
Policy_Billing_Code                         424431 non-null object
Policy_Method_Of_Payment                    424431 non-null object
Policy_Reinstatement_Fee_Indicator          424431 non-null object
Policy_Zip_Code_Garaging_Location           424431 non-null object
Vehicle_Territory                           424431 non-null int64
Vehicle_Make_Year                           424431 non-null int64
Vehicle_Make_Description                    424431 non-null object
Vehicle_Performance                         424431 non-null object
Vehicle_New_Cost_Amount                     424431 non-null int64
Vehicle_Symbol                              424431 non-null int64
Vehicle_Number_Of_Driver

In [85]:
#autoIns.drop(['Vehicle_Comprehensive_Coverage_Limit','Vehicle_Number_Of_Drivers_Assigned','PolicyNo','Policy_Zip_Code_Garaging_Location','EEA_Policy_Zip_Code_3', 'Vehicle_Make_Description', 'Vehicle_New_Cost_Amount', 'EEA_PolicyYear','Claim_Count','Frequency','Severity','Loss_Ratio'],axis=1,inplace=True)


'Vehicle_Usage'
'Vehicle_Bodily_Injury_Limit'
'Vehicle_Youthful_Driver_Training_Code'
'EEA_Agency_Type'
'EEA_Prior_Bodily_Injury_Limit’


In [86]:
autoIns = autoIns[['Loss_Amount','Annual_Premium','Vehicle_Physical_Damage_Limit','Vehicle_Med_Pay_Limit','Vehicle_Annual_Miles','Vehicle_Miles_To_Work','Vehicle_Days_Per_Week_Driven','Vehicle_Symbol','Vehicle_Bodily_Injury_Limit','EEA_Prior_Bodily_Injury_Limit','Vehicle_Usage','Vehicle_Passive_Restraint','Vehicle_Anti_Theft_Device']]

In [87]:
autoIns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 424431 entries, 0 to 424430
Data columns (total 13 columns):
Loss_Amount                      424431 non-null float64
Annual_Premium                   424431 non-null float64
Vehicle_Physical_Damage_Limit    424431 non-null int64
Vehicle_Med_Pay_Limit            424431 non-null int64
Vehicle_Annual_Miles             424431 non-null object
Vehicle_Miles_To_Work            424431 non-null int64
Vehicle_Days_Per_Week_Driven     424431 non-null int64
Vehicle_Symbol                   424431 non-null int64
Vehicle_Bodily_Injury_Limit      407105 non-null object
EEA_Prior_Bodily_Injury_Limit    407105 non-null object
Vehicle_Usage                    424431 non-null object
Vehicle_Passive_Restraint        424431 non-null object
Vehicle_Anti_Theft_Device        424431 non-null object
dtypes: float64(2), int64(5), object(6)
memory usage: 42.1+ MB


autoIns = autoIns[autoIns['EEA_Policy_Zip_Code_3']!='Unknown']

autoIns.shape

In [88]:
autoIns = autoIns[autoIns['Vehicle_Bodily_Injury_Limit']!='1M-1M  ']

In [89]:
autoIns.shape

(424389, 13)

In [90]:
autoIns.shape

(424389, 13)

PolicyNo                                    424431 non-null int64       
Policy_Company                              424431 non-null object
    One-Hot Encoding
    
    Standard
    Preferred
    
Policy_Installment_Term                     424431 non-null int64
Policy_Billing_Code                         424431 non-null object
    One-Hot Encoding
    
    Direct Billed To Insured
    Premium Finance
    
Policy_Method_Of_Payment                    424431 non-null object
    One-Hot Encoding
    
    Pre-Paid
    Installment

Policy_Reinstatement_Fee_Indicator          424431 non-null object
    One-Hot Encoding
    
    N
    Y

*Policy_Zip_Code_Garaging_Location           424431 non-null object(becz "EEA_Policy_Zip_Code_3" is present)

Vehicle_Territory                           424431 non-null int64
Vehicle_Make_Year                           424431 non-null int64
*Vehicle_Make_Description                    424431 non-null object (becz "Vehicle_Performance" is present)
Vehicle_Performance                         424431 non-null object
    One-Hot Encoding
    
    High
    Intermediate
    Sports
    Sports Premium
    Standard
    
*Vehicle_New_Cost_Amount                     424431 non-null int64
    Random replacement of -1 or can be discarded (As -1 has 417345 Records)
    
Vehicle_Symbol                              424431 non-null int64
    (-1,1,2 replaced with 3...as range is 3 to 27)
    
Vehicle_Number_Of_Drivers_Assigned          424431 non-null int64

Vehicle_Usage                               424431 non-null object
    One-Hot Encoding
    
    Business
    Farm
    Pleasure
    Work
    
Vehicle_Miles_To_Work                       424431 non-null int64
    
    (Random Value Generation)
     https://news.leavitt.com/personal-insurance/vehicle-use-and-why-it-affects-auto-insurance/
        
Vehicle_Days_Per_Week_Driven                424431 non-null int64
    (Replace -1, 8,9 with 5 as 5 has majority of records)
    
Vehicle_Annual_Miles                        424431 non-null object
    Calculation : (Vehicle_Miles_To_Work*Vehicle_Days_Per_Week_Driven*52)
     (Replace 'unknown' with 0s )   
    
Vehicle_Anti_Theft_Device                   424431 non-null object
    (KNN imputation of 'unknown' and One hot encoding afterwards)
    
    Active Disabling
    Alarm Only
    Not applicable
    Passive Disabling - Vehicle Recovery
    
    
Vehicle_Passive_Restraint                   424431 non-null int64
    (KNN imputation of 'unknown' and One hot encoding afterwards)
    
    Y
    N
    
Vehicle_Age_In_Years                        424431 non-null int64
    (Not Related To any)
    
Vehicle_Med_Pay_Limit                       424431 non-null int64
(Replace -1 with 0)
    
Vehicle_Bodily_Injury_Limit                 407105 non-null object
    (Ranges : 100-300(130255), (3)100-500(502), 1M-1M(42-Drop), (1)25-50(145145), 250-500(21757), 300-500(617), (2)50-100(107950), 500-500(837), NA-replace with 0-0)
    Do one-hot encoding afterwards
    
    0-0(NA-17326)
    25-50(145145)
    50-100(107950)
    100-500(130255+502+21757+617+837 = 153968)
    
Vehicle_Physical_Damage_Limit               424431 non-null int64
    (Replace -1 with 0s )
    
Vehicle_Comprehensive_Coverage_Indicator    424431 non-null object
    One-Hot Encoding
    
    N
    Y
    
Vehicle_Comprehensive_Coverage_Limit        424431 non-null int64
    (Replace -1 with 0s )
    
Note : Vehicle_Bodily_Injury_Limit, Vehicle_Physical_Damage_Limit, Vehicle_Comprehensive_Coverage_Indicator and Vehicle_Comprehensive_Coverage_Limit are related
    
Vehicle_Collision_Coverage_Indicator        424431 non-null object
    One-Hot Encoding
    
    N
    Y
    
Vehicle_Collision_Coverage_Deductible       424431 non-null int64
    (Replace -1 with 0s )

Note :  Vehicle_Collision_Coverage_Indicator and  Vehicle_Collision_Coverage_Deductible are related
    
Driver_Total                                424431 non-null int64

Driver_Total_Male                           424431 non-null int64
Driver_Total_Female                         424431 non-null int64
Driver_Total_Single                         424431 non-null int64
Driver_Total_Married                        424431 non-null int64
Driver_Total_Related_To_Insured_Self        424431 non-null int64
Driver_Total_Related_To_Insured_Spouse      424431 non-null int64
Driver_Total_Related_To_Insured_Child       424431 non-null int64
Driver_Total_Licensed_In_State              424431 non-null int64
Driver_Minimum_Age                          424431 non-null int64
Driver_Maximum_Age                          424431 non-null int64
Driver_Total_Teenager_Age_15_19             424431 non-null int64
Driver_Total_College_Ages_20_23             424431 non-null int64
Driver_Total_Young_Adult_Ages_24_29         424431 non-null int64
Driver_Total_Low_Middle_Adult_Ages_30_39    424431 non-null int64
Driver_Total_Middle_Adult_Ages_40_49        424431 non-null int64
Driver_Total_Adult_Ages_50_64               424431 non-null int64
Driver_Total_Senior_Ages_65_69              424431 non-null int64
Driver_Total_Upper_Senior_Ages_70_plus      424431 non-null int64

Vehicle_Youthful_Driver_Indicator           424431 non-null object
    One-Hot Encoding
    
    N
    Y
    
Vehicle_Youthful_Driver_Training_Code       424431 non-null object
    One-Hot Encoding
    
    Not Applicable
    With Driver Training
    With or Without Driver Training
    Without Driver Training
    
Vehicle_Youthful_Good_Student_Code          424431 non-null object
   One-Hot Encoding
    
    Eligible for good student credit
    Not Eligible for good student credit
    
Vehicle_Driver_Points                       424431 non-null int64
Vehicle_Safe_Driver_Discount_Indicator      424431 non-null object
   One-Hot Encoding
    
    Y
    N
    
EEA_Liability_Coverage_Only_Indicator       424431 non-null object
   One-Hot Encoding
    
    Y
    N
    
EEA_Multi_Auto_Policies_Indicator           424431 non-null object
   One-Hot Encoding
    
    Y
    N
    
EEA_Policy_Zip_Code_3                       424431 non-null object
    (Remove 'Unknown' Records(total 3468))
    
?EEA_Policy_Tenure                           424431 non-null float64
    (Replace -1 with 0s)
    
EEA_Agency_Type                             424431 non-null object
   One-Hot Encoding
    
    Standard
    Preferred
    Non-Standard
    Hybrid
    
EEA_Packaged_Policy_Indicator               424431 non-null object
   One-Hot Encoding
    
    Y
    N
    
EEA_Full_Coverage_Indicator                 424431 non-null object
   One-Hot Encoding
    
    Y
    N
    
EEA_Prior_Bodily_Injury_Limit               407105 non-null object
    (Ranges : (3)100-400(21757), (3)100-200(130255), (2)20-50(145145), (3)200-400(617), (3)300-300(837), (2)40-100(107950), (3)75-300(502), (3)750-750(42),(1)NA-replace with 0-0(17326))
    Do one-hot encoding afterwards
    
    0-0(NA-17326)
    20-100(145145+107950=253095)
    100-750(21757+130255+617+837+502+42 = 154010)    
    
*EEA_PolicyYear                              424431 non-null int64(can be dropped as data for only single year 2006 is present)

SYS_Renewed                                 424431 non-null object
   One-Hot Encoding
    
    Y
    N
    
SYS_New_Business                            424431 non-null object
   One-Hot Encoding
    
    Y
    N
    
Annual_Premium                              424431 non-null float64
    
Claim_Count                                 424431 non-null int64
Loss_Amount                                 424431 non-null float64(y)
Frequency                                   424431 non-null float64
Severity                                    424431 non-null float64
Loss_Ratio                                  424431 non-null float64
dtypes: float64(6), int64(37), object(26)
memory usage: 226.7+ MB#

In [91]:
def rangee(x):
    if 'nan' in str(x):
        x = '0-0'
    elif x == '100-300':
        x = '100-500'
    elif x == '100-500':
        x = '100-500'
    elif x == '250-500':
        x = '100-500'
    elif x == '300-500':
        x = '100-500'
    elif x == '500-500':
        x = '100-500'
    return x

autoIns.Vehicle_Bodily_Injury_Limit = autoIns.Vehicle_Bodily_Injury_Limit.apply(rangee)

In [92]:
autoIns.Vehicle_Bodily_Injury_Limit.unique()

array(['100-500', '50-100 ', '25-50  ', '0-0'], dtype=object)

In [93]:
def rangee(x):
    if 'nan' in str(x):
        x = '0-0'
    elif x == '20-50':
        x = '20-100'
    elif x == '40-100 ':
        x = '20-100'
    elif x == '100-400':
        x = '100-750'
    elif x == '100-200':
        x = '100-750'
    elif x == '75-300':
        x = '100-750'
    elif x == '200-400':
        x = '100-750'
    elif x == '300-300':
        x = '100-750'
    elif x == '750-750':
        x = '100-750'
    return x

autoIns.EEA_Prior_Bodily_Injury_Limit = autoIns.EEA_Prior_Bodily_Injury_Limit.apply(rangee)

In [94]:
autoIns.EEA_Prior_Bodily_Injury_Limit.unique()

array(['100-750', '20-100', '0-0'], dtype=object)

In [95]:
autoIns.shape

(424389, 13)

In [96]:
cols=['Vehicle_Usage','Vehicle_Bodily_Injury_Limit','EEA_Prior_Bodily_Injury_Limit']
autoIns=pd.get_dummies(data=autoIns, columns=cols)

In [97]:
autoIns.shape

(424389, 21)

In [98]:
#autoIns['Policy_Company']=autoIns['Policy_Company'].map({'Standard ': 1,'Preferred': 0})
#autoIns['Policy_Reinstatement_Fee_Indicator']=autoIns['Policy_Reinstatement_Fee_Indicator'].map({'Y': 1,'N': 0})
#autoIns['Vehicle_Comprehensive_Coverage_Indicator']=autoIns['Vehicle_Comprehensive_Coverage_Indicator'].map({'Y': 1,'N': 0})
#autoIns['Vehicle_Collision_Coverage_Indicator']=autoIns['Vehicle_Collision_Coverage_Indicator'].map({'Y': 1,'N': 0})
#autoIns['Vehicle_Youthful_Driver_Indicator']=autoIns['Vehicle_Youthful_Driver_Indicator'].map({'Y': 1,'N': 0})
#autoIns['Vehicle_Safe_Driver_Discount_Indicator']=autoIns['Vehicle_Safe_Driver_Discount_Indicator'].map({'Y': 1,'N': 0})
#autoIns['EEA_Liability_Coverage_Only_Indicator']=autoIns['EEA_Liability_Coverage_Only_Indicator'].map({'Y': 1,'N': 0})
#autoIns['EEA_Multi_Auto_Policies_Indicator']=autoIns['EEA_Multi_Auto_Policies_Indicator'].map({'Y': 1,'N': 0})
#autoIns['EEA_Packaged_Policy_Indicator']=autoIns['EEA_Packaged_Policy_Indicator'].map({'Y': 1,'N': 0})
#autoIns['EEA_Full_Coverage_Indicator']=autoIns['EEA_Full_Coverage_Indicator'].map({'Y': 1,'N': 0})
#autoIns['SYS_Renewed']=autoIns['SYS_Renewed'].map({'Y': 1,'N': 0})
#autoIns['SYS_New_Business']=autoIns['SYS_New_Business'].map({'Y': 1,'N': 0})


In [99]:
#autoIns['Policy_Billing_Code']=autoIns['Policy_Billing_Code'].map({'Direct Billed to Insured': 1,'Premium Finance': 0})

In [100]:
autoIns.shape

(424389, 21)

In [101]:
#autoIns.Policy_Billing_Code.unique()

In [102]:
#autoIns['Policy_Method_Of_Payment']=autoIns['Policy_Method_Of_Payment'].map({'Pre-paid': 1,'Installment': 0})
#autoIns['Vehicle_Youthful_Good_Student_Code']=autoIns['Vehicle_Youthful_Good_Student_Code'].map({'Eligible for Good Student Credit    ': 1,'Not Eligible for Good Student Credit': 0})

scale_mapper = {'Standard':5, 
                'Intermediate':4,
                'High':3,
                'Sports':2,
                'Sports Premium':1}
autoIns['Vehicle_Performance'] = autoIns['Vehicle_Performance'].replace(scale_mapper)


In [103]:
#autoIns['Vehicle_Performance'].value_counts()

In [104]:
knn_df = autoIns[(autoIns["Vehicle_Passive_Restraint"]!='Unknown')]
X= knn_df[['Vehicle_Symbol', 'Annual_Premium']]
y = knn_df[['Vehicle_Passive_Restraint']]

In [105]:
X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [106]:
X_train.shape

(339446, 2)

In [107]:
neigh = KNeighborsClassifier(n_neighbors = 20, weights='distance', algorithm='auto')
neigh.fit(X_train, Y_train) 
y_pred = neigh.predict(X_test)
print(y_pred)

['Y' 'N' 'Y' ... 'Y' 'Y' 'Y']


In [108]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_test,y_pred)

0.7983903278263533

In [109]:
df11 = autoIns[(autoIns.Vehicle_Passive_Restraint == "Unknown")]

In [110]:
df11

Loss_Amount  Annual_Premium  Vehicle_Physical_Damage_Limit  \
16572           0.0          242.74                          25000   
16573           0.0          252.28                          25000   
27936           0.0          369.94                             -1   
27937           0.0          340.26                             -1   
28025           0.0          483.36                          25000   
32053           0.0          490.78                          50000   
32699           0.0          263.94                          50000   
32700           0.0          273.48                          50000   
39805           0.0          270.30                         100000   
48997           0.0          415.52                          25000   
48998           0.0          415.52                          25000   
55803           0.0          440.96                         100000   
57752           0.0          218.36                          50000   
57753           0.0          200.34                          50000   
66166           0.0          339.20                         100000   
66167           0.0          338.14                         100000   
74935           0.0          124.02                          25000   
75378           0.0          309.52                          25000   
77994           0.0          516.22                          25000   
89837           0.0          170.66                          50000   
89838           0.0          169.60                          50000   
90345           0.0          380.54                             -1   
90766           0.0          553.32                         100000   
92558           0.0          389.02                          25000   
92559           0.0          369.94                          25000   
102931          0.0          334.96                          25000   
105650          0.0          253.34                          35000   
105651          0.0          214.12                          35000   
105883          0.0          294.68                          50000   
105884          0.0          292.56                          50000   
...             ...             ...                            ...   
197403          0.0          644.48                          50000   
198816          0.0          271.36                          25000   
203466          0.0          366.76                          50000   
203971          0.0          306.34                          50000   
209600          0.0          414.46                          50000   
213382          0.0          143.10                          50000   
221633          0.0          351.92                          25000   
221634          0.0          350.86                          25000   
254412          0.0           68.89                          50000   
263490          0.0          347.68                          50000   
267347          0.0          521.52                          25000   
268176          0.0          177.88                          50000   
274472          0.0          169.60                          25000   
274473          0.0           51.68                          25000   
274474          0.0          191.86                          25000   
293365          0.0          157.94                          50000   
298069          0.0          298.13                          25000   
311115          0.0          523.64                          50000   
316757          0.0          196.10                          35000   
318132          0.0          665.68                         100000   
325599          0.0          190.80                          50000   
339306          0.0          491.84                          25000   
346290          0.0          583.00                          25000   
350892          0.0          171.72                          50000   
354427          0.0          144.16                          50000   
370112          0.

In [111]:
neigh = KNeighborsClassifier(n_neighbors = 20, weights='distance', algorithm='auto')
neigh.fit(X_train, Y_train) 
y_pred = neigh.predict(df11[['Vehicle_Symbol', 'Annual_Premium']])
print(y_pred)

['Y' 'N' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y'
 'Y' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'Y' 'Y'
 'N' 'Y' 'Y' 'Y' 'N' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y' 'Y'
 'Y' 'Y' 'Y' 'Y' 'Y' 'N' 'Y' 'Y' 'N' 'Y' 'N' 'N' 'N' 'N' 'Y' 'N' 'Y' 'N'
 'Y' 'Y' 'N' 'N' 'Y' 'Y' 'Y' 'Y' 'N']


In [112]:
autoIns.Vehicle_Passive_Restraint.value_counts()

Y          311904
N          112404
Unknown        81
Name: Vehicle_Passive_Restraint, dtype: int64

In [113]:
autoIns.loc[autoIns.Vehicle_Passive_Restraint == 'Unknown', 'Vehicle_Passive_Restraint'] = y_pred

In [114]:
autoIns.Vehicle_Passive_Restraint.value_counts()

Y    311965
N    112424
Name: Vehicle_Passive_Restraint, dtype: int64

In [115]:
autoIns['Vehicle_Passive_Restraint']=autoIns['Vehicle_Passive_Restraint'].map({'Y': 1,'N': 0})

In [116]:
autoIns.Vehicle_Passive_Restraint.value_counts()

1    311965
0    112424
Name: Vehicle_Passive_Restraint, dtype: int64

In [117]:
tft1 = autoIns[(autoIns["Vehicle_Anti_Theft_Device"]!="Unknown")]
df11 = autoIns[(autoIns.Vehicle_Anti_Theft_Device == "Unknown")]
X = tft1[['Vehicle_Symbol']]
y = tft1[['Vehicle_Anti_Theft_Device']]

In [118]:
X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [119]:
neigh = KNeighborsClassifier(n_neighbors = 18, weights='distance', algorithm='auto')
neigh.fit(X_train, Y_train) 
y_pred = neigh.predict(df11[['Vehicle_Symbol']])
print(y_pred)

['Passive Disabling-Vehicle Recovery' 'Not Applicable' 'Not Applicable'
 'Not Applicable' 'Not Applicable']


In [120]:
autoIns.loc[autoIns.Vehicle_Anti_Theft_Device == 'Unknown', 'Vehicle_Anti_Theft_Device'] = y_pred

In [121]:
autoIns.Vehicle_Anti_Theft_Device.value_counts()

Not Applicable                        248620
Passive Disabling-Vehicle Recovery    133348
Alarm Only                             31843
Active Disabling                       10578
Name: Vehicle_Anti_Theft_Device, dtype: int64

In [122]:
cols=['Vehicle_Anti_Theft_Device']
autoIns=pd.get_dummies(data=autoIns, columns=cols)

In [123]:
autoIns.Vehicle_Symbol.value_counts()

 8     48117
 11    46361
 10    45186
 12    45128
 14    36730
 13    36180
 7     29845
 15    28348
 6     19994
 16    18638
 5     16632
 17    12518
 4     10099
 18     8545
 19     5177
 3      3889
 20     3543
 21     2653
 2      2071
 22     1696
 23     1110
 1       681
 24      616
 25      275
 26      203
 27      120
-1        25
 9         9
Name: Vehicle_Symbol, dtype: int64

In [124]:
autoIns.loc[autoIns['Vehicle_Symbol'] < 3, 'Vehicle_Symbol'] = 3

In [125]:
autoIns.Vehicle_Symbol.value_counts()

8     48117
11    46361
10    45186
12    45128
14    36730
13    36180
7     29845
15    28348
6     19994
16    18638
5     16632
17    12518
4     10099
18     8545
3      6666
19     5177
20     3543
21     2653
22     1696
23     1110
24      616
25      275
26      203
27      120
9         9
Name: Vehicle_Symbol, dtype: int64

In [126]:
autoIns.Vehicle_Days_Per_Week_Driven.value_counts()

 5    424236
 0        63
-1        19
 1        17
 3        16
 6        15
 2         7
 7         5
 4         5
 8         4
 9         2
Name: Vehicle_Days_Per_Week_Driven, dtype: int64

In [127]:
autoIns.loc[((autoIns['Vehicle_Days_Per_Week_Driven'] == -1) | (autoIns['Vehicle_Days_Per_Week_Driven'] > 7)),'Vehicle_Days_Per_Week_Driven'] = 5

In [128]:
autoIns.Vehicle_Days_Per_Week_Driven.value_counts()

5    424261
0        63
1        17
3        16
6        15
2         7
7         5
4         5
Name: Vehicle_Days_Per_Week_Driven, dtype: int64

In [129]:
autoIns.Vehicle_Miles_To_Work.unique()

array([-1, 10, 12,  7, 21, 14,  8,  4,  2, 13,  6,  5, 25, 15,  3,  1, 42,
       20, 35, 40, 30, 50, 11,  9, 45, 22, 16, 26, 17, 23, 18,  0, 24, 27,
       60, 31, 29, 32, 99, 28, 38, 49, 19, 37, 48, 34, 36, 95, 39, 62, 43,
       47, 52, 46, 65, 90, 55, 80, 58, 44, 51, 57, 70, 71, 56, 33, 83, 75,
       61, 64, 41, 93, 54, 63, 53, 78, 67, 72, 98, 59])

In [130]:
autoIns.Vehicle_Miles_To_Work.value_counts()

-1     266197
 10     30906
 5      18531
 12     12986
 8      11489
 6       9314
 3       8511
 14      8279
 4       7378
 2       7086
 7       5655
 20      5253
 1       4148
 13      3641
 25      2879
 30      2768
 11      2539
 9       2538
 15      2254
 50      1522
 18      1415
 16      1201
 0       1064
 40       999
 35       940
 17       700
 22       615
 45       434
 23       388
 24       334
        ...  
 49        23
 55        23
 39        21
 65        17
 80        16
 52        15
 75        14
 41        13
 46        13
 44        10
 62         8
 90         7
 53         6
 58         4
 51         4
 57         4
 56         4
 63         3
 54         3
 64         2
 72         2
 78         2
 83         2
 95         2
 59         1
 67         1
 71         1
 61         1
 93         1
 98         1
Name: Vehicle_Miles_To_Work, Length: 80, dtype: int64

In [131]:
def replace_with_random(x):        
    try:        
        if x == -1:   
            while(True):
                rand_num = random.choice(autoIns.Vehicle_Miles_To_Work)
                if rand_num != -1:
                    return rand_num    
                else:
                    continue
        else:            
            return x        
    except Exception as e:
        return 0
    
autoIns["Vehicle_Miles_To_Work"] = autoIns["Vehicle_Miles_To_Work"].apply(replace_with_random)

In [132]:
autoIns.Vehicle_Miles_To_Work.value_counts()

10    83030
5     49658
12    34808
8     30962
6     24789
3     22986
14    22187
4     19768
2     18888
7     15197
20    14204
1     11096
13     9733
25     7651
30     7420
11     6862
9      6748
15     6058
50     4087
18     3771
16     3226
0      2961
40     2684
35     2553
17     1931
22     1620
45     1130
23     1070
24      887
28      808
      ...  
49       62
39       61
55       53
80       45
52       44
65       41
46       40
41       33
75       32
44       31
53       22
62       19
90       14
56       13
58       12
57       12
51       12
63        9
78        7
54        7
98        6
95        6
83        5
67        4
72        4
64        4
61        3
59        2
71        2
93        2
Name: Vehicle_Miles_To_Work, Length: 79, dtype: int64

In [133]:
autoIns.Vehicle_Annual_Miles.unique()

array(['Unknown', '0', '30000', '50000', '20000', '70000', '100000',
       '500000', '60000', '150000'], dtype=object)

In [134]:
autoIns['Vehicle_Annual_Miles'] = autoIns['Vehicle_Miles_To_Work'] * autoIns['Vehicle_Days_Per_Week_Driven'] * 52

In [135]:
autoIns.Vehicle_Annual_Miles.value_counts()

2600     83012
1300     49640
3120     34804
2080     30953
1560     24782
780      22982
3640     22187
1040     19757
520      18890
1820     15193
5200     14200
260      11092
3380      9732
6500      7651
7800      7418
2860      6860
2340      6747
3900      6057
13000     4085
4680      3771
4160      3226
0         3023
10400     2682
9100      2553
4420      1931
5720      1620
11700     1130
5980      1070
6240       888
7280       807
         ...  
13260       12
16380        9
14040        7
20280        7
24700        6
25480        6
21580        5
16640        4
17420        4
18720        4
156          3
208          3
624          3
15860        3
2548         2
4368         2
15340        2
1872         2
936          2
18460        2
24180        2
312          2
2496         2
2912         2
3744         2
15444        1
572          1
676          1
1456         1
1716         1
Name: Vehicle_Annual_Miles, Length: 95, dtype: int64

In [136]:
autoIns.Vehicle_Med_Pay_Limit.unique()

array([ 1000,  2000, 50000,  5000,    -1, 10000, 25000])

In [137]:
autoIns.loc[autoIns['Vehicle_Med_Pay_Limit'] == -1,'Vehicle_Med_Pay_Limit'] = 0

In [138]:
autoIns.Vehicle_Med_Pay_Limit.unique()

array([ 1000,  2000, 50000,  5000,     0, 10000, 25000])

In [139]:
autoIns.Vehicle_Physical_Damage_Limit.unique()

array([100000,  50000,  35000,  25000,     -1, 250000, 500000])

In [140]:
autoIns.loc[autoIns['Vehicle_Physical_Damage_Limit'] == -1,'Vehicle_Physical_Damage_Limit'] = 0

In [141]:
autoIns.Vehicle_Physical_Damage_Limit.unique()

array([100000,  50000,  35000,  25000,      0, 250000, 500000])

autoIns.Vehicle_Comprehensive_Coverage_Limit.unique()

autoIns.loc[autoIns['Vehicle_Comprehensive_Coverage_Limit'] == -1,'Vehicle_Comprehensive_Coverage_Limit'] = 0

In [142]:
#autoIns.Policy_Company.unique()

In [143]:
#autoIns.Policy_Company.value_counts()

In [144]:
autoIns.head()

Loss_Amount  Annual_Premium  Vehicle_Physical_Damage_Limit  \
0          0.0          320.12                         100000   
1          0.0          259.70                          50000   
2          0.0          613.74                         100000   
3          0.0          541.66                          35000   
4          0.0          541.66                          35000   

   Vehicle_Med_Pay_Limit  Vehicle_Annual_Miles  Vehicle_Miles_To_Work  \
0                   1000                   780                      3   
1                   2000                  2080                      8   
2                  50000                  2080                      8   
3                   5000                  1820                      7   
4                   5000                  1300                      5   

   Vehicle_Days_Per_Week_Driven  Vehicle_Symbol  Vehicle_Passive_Restraint  \
0                             5              11                          1   
1                             5              11                          0   
2                             5              10                          1   
3                             5              12                          1   
4                             5              12                          1   

   Vehicle_Usage_Business  ...  Vehicle_Bodily_Injury_Limit_100-500  \
0                       0  ...                                    1   
1                       0  ...                                    1   
2                       0  ...                                    1   
3                       0  ...                                    0   
4                       0  ...                                    0   

   Vehicle_Bodily_Injury_Limit_25-50    Vehicle_Bodily_Injury_Limit_50-100   \
0                                    0                                    0   
1                                    0                                    0   
2                                    0                                    0   
3                                    0                                    1   
4                                    0                                    1   

   EEA_Prior_Bodily_Injury_Limit_0-0  EEA_Prior_Bodily_Injury_Limit_100-750  \
0                                  0                                      1   
1                                  0                                      1   
2                                  0                                      1   
3                                  0                                      0   
4                                  0                                      0   

   EEA_Prior_Bodily_Injury_Limit_20-100  \
0                                     0   
1                                     0   
2                                     0   
3                                     1   
4                                     1   

   Vehicle_Anti_Theft_Device_Active Disabling  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

   Vehicle_Anti_Theft_Device_Alarm Only  \
0                                     0   
1                                     0   
2                                     0   
3                                     0   
4                                     0   

   Vehicle_Anti_Theft_Device_Not Applicable  \
0                                         1   
1                                         1   
2                                         0   
3                                         0   
4                                         0   

   Vehicle_Anti_Theft_Device_Passive Disabling-Vehicle Recovery  
0                                                  0             
1                                                  0             
2                                

In [161]:
autoIns.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 424389 entries, 0 to 424430
Data columns (total 24 columns):
Loss_Amount                                                     424389 non-null float64
Annual_Premium                                                  424389 non-null float64
Vehicle_Physical_Damage_Limit                                   424389 non-null int64
Vehicle_Med_Pay_Limit                                           424389 non-null int64
Vehicle_Annual_Miles                                            424389 non-null int64
Vehicle_Miles_To_Work                                           424389 non-null int64
Vehicle_Days_Per_Week_Driven                                    424389 non-null int64
Vehicle_Symbol                                                  424389 non-null int64
Vehicle_Passive_Restraint                                       424389 non-null int64
Vehicle_Usage_Business                                          424389 non-null uint8
Vehicle_Usage_Farm       

In [162]:
X = autoIns.drop('Loss_Amount',axis=1)
y = autoIns['Loss_Amount']

In [163]:
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state=42)

In [164]:
autoIns.Loss_Amount.value_counts()

0.00        405522
54.50         1463
1090.00        260
49.05          212
43.60          102
109.00         101
38.15           74
2180.00         55
5450.00         49
27.25           49
103.55          25
27250.00        22
65.35           21
32.70           18
65.40           16
92.65           16
10900.00        16
163.50          14
98.10           13
54500.00        12
3270.00         11
81.75            9
272.50           9
76.25            8
545.00           7
76.30            7
4360.00          7
1144.50          6
87.20            5
179.85           5
             ...  
2981.22          1
2038.05          1
2079.45          1
727.62           1
1201.23          1
16412.40         1
7063.20          1
15173.52         1
1700.75          1
4066.35          1
420.76           1
1315.62          1
414.46           1
7120.68          1
1896.94          1
13465.32         1
727.13           1
5261.98          1
3791.87          1
2078.95          1
745.32           1
6651.89     

## Linear regression With standardization

In [165]:
scaler = preprocessing.StandardScaler(copy=True, with_mean=True, with_std=True)
x_train_std = scaler.fit_transform(x_train)
x_train_std = pd.DataFrame(x_train_std, columns=x_train.columns)
x_test_std = scaler.transform(x_test)
x_test_std = pd.DataFrame(x_test_std, columns=x_test.columns)

In [150]:
lm = LinearRegression(copy_X=True, fit_intercept=False, n_jobs=1)
lm.fit(x_train_std,y_train)
y_test_pred=lm.predict(x_test_std)
print(lm.score(x_test_std,y_test))
print(metrics.mean_squared_error(y_test,y_test_pred))

-0.004678162472963487
8539950.462910337


from sklearn.preprocessing import PolynomialFeatures
  
poly = PolynomialFeatures(degree = 4)
X_poly = poly.fit_transform(x_train)
#poly.fit(X_poly, y_train)

lm_poly = LinearRegression()
lm_poly.fit(X_poly, y_train)

y_poly_pred = lm_poly.predict(x_poly)

rmse = np.sqrt(mean_squared_error(y_train,y_poly_pred))
r2 = r2_score(y_train,y_poly_pred)
print(rmse)
print(r2)

## Linear regression With Random Forest

from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators=300, random_state=50, n_jobs=-1, oob_score = True, max_features = "auto", min_samples_leaf = 60)
regressor.fit(x_train_std, y_train)
#print(regressor.score(x_test_std,y_test))
y_pred = regressor.predict(x_test_std)

print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

#10
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

#99
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

#50
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

#20
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

## <b>Test</b>

In [169]:
x_test = pd.read_csv("x_test.csv")

In [170]:
temp = x_test.drop('portfolio', axis=1)

KeyError: "['portfolio'] not found in axis"

In [168]:
x_test_standard = scaler.transform(temp)
x_test_standard = pd.DataFrame(x_test_standard, columns=temp.columns)

ValueError: operands could not be broadcast together with shapes (985095,73) (23,) (985095,73) 

In [ ]:
x_test_standard

In [557]:
y_lossAmt_pred = lm.predict(x_test_standard)

In [558]:
#y_lossAmt_pred = regressor.predict(x_test_standard)

In [154]:
y_lossAmt_pred

NameError: name 'y_lossAmt_pred' is not defined

In [155]:
x_test['Loss_Amount'] = 0
x_test.Loss_Amount.value_counts()

0    985095
Name: Loss_Amount, dtype: int64

In [561]:
x_test.loc[x_test.Loss_Amount == 0, 'Loss_Amount'] = y_lossAmt_pred

In [562]:
x_test.Annual_Premium.head()

0    203.52
1    101.76
2    198.22
3    447.32
4    172.78
Name: Annual_Premium, dtype: float64

In [563]:
x_test.Loss_Amount.head()

0   -192.193005
1   -184.193005
2   -238.193005
3     57.806995
4    -66.193005
Name: Loss_Amount, dtype: float64

In [564]:
x_test[x_test.Loss_Amount < 2].Loss_Amount.count()

526514

In [566]:
sum_df = x_test.groupby(['portfolio']).agg({'Annual_Premium': 'sum', 'Loss_Amount': 'sum'})
sum_df['ln_LR'] = np.log(sum_df['Loss_Amount'] / sum_df['Annual_Premium'])

In [567]:
sum_df.reset_index(level=0, inplace=True)
data = sum_df[['portfolio','ln_LR']]
#data.to_csv('Loss_ratio1.txt', header=["ID", "ln_LR"],index = None, sep=',', mode='a')

In [477]:
data['ln_LR'] = data['ln_LR']/10

In [480]:
data.to_csv('LR.txt', header=["ID", "ln_LR"],index = None, sep=',', mode='a')

In [568]:
data

portfolio     ln_LR
0      portfolio_1 -2.999620
1     portfolio_10 -3.440880
2     portfolio_11 -4.737413
3    portfolio_111 -3.538350
4    portfolio_112 -2.950626
5    portfolio_113 -2.812637
6    portfolio_114 -3.639358
7    portfolio_115 -3.740435
8    portfolio_116 -2.738166
9    portfolio_117       NaN
10   portfolio_118 -3.739887
11   portfolio_119 -3.199686
12    portfolio_12 -5.427589
13   portfolio_120 -3.032088
14    portfolio_13 -4.124656
15   portfolio_131 -2.932702
16   portfolio_132 -3.108957
17   portfolio_133 -2.908294
18   portfolio_134 -2.811022
19   portfolio_135 -4.036520
20   portfolio_136 -2.687776
21   portfolio_137 -2.603567
22   portfolio_138 -3.198571
23   portfolio_139 -4.238554
24    portfolio_14 -4.165762
25   portfolio_140 -3.467778
26    portfolio_15 -2.738332
27   portfolio_151 -3.741060
28   portfolio_152 -4.484159
29   portfolio_153 -2.638042
..             ...       ...
300  portfolio_578 -5.951027
301  portfolio_579 -3.975491
302   portfolio_58 -2.727285
303  portfolio_580 -3.564962
304   portfolio_59 -3.966603
305  portfolio_591 -2.453740
306  portfolio_592 -2.783273
307  portfolio_593 -2.528040
308  portfolio_594 -2.758941
309  portfolio_595 -2.618259
310  portfolio_596 -2.624852
311  portfolio_597 -2.791384
312  portfolio_598 -2.523601
313  portfolio_599 -2.480567
314    portfolio_6 -3.467828
315   portfolio_60 -5.219239
316  portfolio_600 -2.557391
317   portfolio_61 -2.666553
318   portfolio_62       NaN
319   portfolio_63 -2.904191
320   portfolio_64 -5.741982
321   portfolio_65 -3.002080
322   portfolio_66 -3.217923
323   portfolio_67 -3.263494
324   portfolio_68 -2.668815
325   portfolio_69 -3.050623
326    portfolio_7 -3.513188
327   portfolio_70 -2.896212
328    portfolio_8 -3.101827
329    portfolio_9 -3.847733

[330 rows x 2 columns]

In [483]:
autoIns.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 424389 entries, 0 to 424430
Data columns (total 74 columns):
Policy_Company                                                           424389 non-null int64
Policy_Installment_Term                                                  424389 non-null int64
Policy_Billing_Code                                                      424389 non-null int64
Policy_Method_Of_Payment                                                 424389 non-null int64
Policy_Reinstatement_Fee_Indicator                                       424389 non-null int64
Vehicle_Territory                                                        424389 non-null int64
Vehicle_Make_Year                                                        424389 non-null int64
Vehicle_Performance                                                      424389 non-null int64
Vehicle_Symbol                                                           424389 non-null int64
Vehicle_Miles_To_Work             